### ISSUE TITLE
Take advantage of the `StrictTVar` invariant to check for NF

### ISSUE THREAD

**Issue by edsko** [input-output-hk/ouroboros-network#969]
See https://github.com/input-output-hk/ouroboros-network/pull/965#discussion_r318893225 .

**Comment by edsko** [comment]
Some problems: `ChainFragment` uses `FingerTree` under the hood, but that is not strict. I'm not entirely sure if we can write a strict version; [the paper](http://www.staff.city.ac.uk/~ross/papers/FingerTree.pdf) says that laziness is "essential", but I'm not sure in which way. (Doesn't `Data.Map` use a similar technique..? But _is_ strict? Needs investigation.) Related: https://github.com/input-output-hk/ouroboros-network/pull/853

Another comment: the DB is storing cloures, and @intricate tells me that closures are checked for thunks also; this might be hard to track down, perhaps.

He advices:

> Unfortunately I haven't got a great visual tool for helping out with this but we did add Cardano.Prelude.GHC.Heap.Tree which allows you to render the heap object tree of any given Haskell value: https://github.com/input-output-hk/cardano-prelude/blob/master/src/Cardano/Prelude/GHC/Heap/Tree.hs
See the functions buildClosureTree and buildAndRenderClosureTree. What I'll usually do is build and render/print the tree and then search for thunk closures in it. Example:
```
let opts = ClosureTreeOptions AnyDepth NoTraverseCyclicClosures
    tree = buildClosureTree opts haskellVal
```

**Comment by edsko** [comment]
FingerTree problem should be addressed by https://github.com/input-output-hk/cardano-prelude/pull/67.

**Comment by mrBliss** [comment]
Addressed by #1062.
